refactor(tofs): avoid using “salt['config.get']” for formula writers

baby-gnu opened the pull request:
We can hide the call to “salt['config.get']” in the macro by only
asking for a namespace where to lookup the “source_files”.

* docs/TOFS_pattern.rst (Example): document the use of the new
  “namespace” parameter.

* template/macros.jinja: add a new “namespace” parameter.
  Lookup files override under
  “<tplroot>:tofs:sources_files:<namespace>” and fallback to
  “source_files” parameter.

* template/config/file.sls (template-config-file-file-managed): use
  the new “namespace” parameter

baby-gnu commented:
Hello.

I think I can rewrite the commit message since it actually breaks things. Maybe there is a better way of doing it, it's open to suggestion.

Regards

myii commented:
@baby-gnu Just had a brief glance -- this looks like a very good idea.  What break things -- just the commit message or something else?

myii commented on template/macros.jinja:
```diff
@@ -52,6 +52,11 @@
       tplroot ~ ':tofs:files_switch',
       default_files_switch
   ) %}
+  {# Lookup files or fallback to source_files parameter #}
```
Please use `{#-` instead of `{#`, so that it doesn't end up rendering a blank line:

```sls
    - source:

      - salt://template/files/ABC/example.tmpl
      - salt://template/files/ABC/example.tmpl.jinja
      - salt://template/files/Debian/example.tmpl
      - salt://template/files/Debian/example.tmpl.jinja
      - salt://template/files/default/example.tmpl
      - salt://template/files/default/example.tmpl.jinja
```

myii replied:
Thanks.

myii commented on template/macros.jinja:
```diff
@@ -66,7 +71,7 @@
     {%- endif %}
     {%- for fs in fsl %}
-      {%- for source_file in source_files %}
+      {%- for source_file in files %}
```
For consistency: `source_file` => `file`.  What do you think?

myii replied:
If doing that, then change below: `source_file.lstrip('/')` => `file.lstrip('/')`.

myii replied:
Update: Forget these comments, I'd prefer to solve [the comment above](https://github.com/saltstack-formulas/template-formula/pull/77/files#r264497818) instead.

myii replied:
Suggestion: `file` => `src_file`, in both places.

myii replied:
Thanks.

myii commented:
@baby-gnu I've had a chance to look at this and test it out and I definitely agree that this is the right way to go forwards -- thanks for the contribution!  However, I'd just like a couple of things cleaned up, as I've mentioned in my comments above.  I also want to ask @aboe76 to have a look, so that we can be confident that we're heading in the right direction.

I've tested before and after with the following in my pillar (includes some extra tests for `files_switch`:

```yaml
  tofs:
    files_switch:
      - any/path/can/be/used/here
      - id
      - osfinger
      - os
      - os_family
    path_prefix: template_alt
    dirs:
      files: files_alt
      default: default_alt
    source_files:
      template-config-file-file-managed:
        - 'example_alt_source.tmpl'
        - 'example_alt.tmpl.jinja'
      formula.autofs.config.master:
        - 'auto_alt.master'
      formula.autofs.config.home:
        - 'auto_alt.homeLdap'
      formula.autofs.config.away:
        - 'auto_alt.awayLdap'
```

I've got the same results before and after:

```sls
    - source:
      - salt://template_alt/files_alt/any/path/can/be/used/here/example_alt_source.tmpl
      - salt://template_alt/files_alt/any/path/can/be/used/here/example_alt.tmpl.jinja
      - salt://template_alt/files_alt/ABC/example_alt_source.tmpl
      - salt://template_alt/files_alt/ABC/example_alt.tmpl.jinja
      - salt://template_alt/files_alt/Ubuntu-16.04/example_alt_source.tmpl
      - salt://template_alt/files_alt/Ubuntu-16.04/example_alt.tmpl.jinja
      - salt://template_alt/files_alt/Ubuntu/example_alt_source.tmpl
      - salt://template_alt/files_alt/Ubuntu/example_alt.tmpl.jinja
      - salt://template_alt/files_alt/Debian/example_alt_source.tmpl
      - salt://template_alt/files_alt/Debian/example_alt.tmpl.jinja
      - salt://template_alt/files_alt/default_alt/example_alt_source.tmpl
      - salt://template_alt/files_alt/default_alt/example_alt.tmpl.jinja

    - source:
      - salt://template_alt/files_alt/any/path/can/be/used/here/auto_alt.master
      - salt://template_alt/files_alt/ABC/auto_alt.master
      - salt://template_alt/files_alt/Ubuntu-16.04/auto_alt.master
      - salt://template_alt/files_alt/Ubuntu/auto_alt.master
      - salt://template_alt/files_alt/Debian/auto_alt.master
      - salt://template_alt/files_alt/default_alt/auto_alt.master

    - source:
      - salt://template_alt/files_alt/any/path/can/be/used/here/auto_alt.homeLdap
      - salt://template_alt/files_alt/ABC/auto_alt.homeLdap
      - salt://template_alt/files_alt/Ubuntu-16.04/auto_alt.homeLdap
      - salt://template_alt/files_alt/Ubuntu/auto_alt.homeLdap
      - salt://template_alt/files_alt/Debian/auto_alt.homeLdap
      - salt://template_alt/files_alt/default_alt/auto_alt.homeLdap

    - source:
      - salt://template_alt/files_alt/any/path/can/be/used/here/auto_alt.awayLdap
      - salt://template_alt/files_alt/ABC/auto_alt.awayLdap
      - salt://template_alt/files_alt/Ubuntu-16.04/auto_alt.awayLdap
      - salt://template_alt/files_alt/Ubuntu/auto_alt.awayLdap
      - salt://template_alt/files_alt/Debian/auto_alt.awayLdap
      - salt://template_alt/files_alt/default_alt/auto_alt.awayLdap
```

So I'm happy with this implementation.

myii commented on template/macros.jinja:
```diff
@@ -52,6 +52,11 @@
       default_files_switch
   ) %}
+  {# Lookup files or fallback to source_files parameter #}
+  {%- set files = salt['config.get'](
```
Suggestion: `files` => `src_files`.

myii replied:
Thanks.

myii commented on template/config/file.sls:
```diff
@@ -13,12 +13,10 @@ include:
-                    )
-              ) }}
+    - source: {{ files_switch(namespace='template-config-file-file-managed',
+                              source_files=['example.tmpl', 'example.tmpl.jinja']
```
Using `namespace=` and `source_files=` seems to make sense at this stage.  However, I think it just makes this longer than in needs to be.  Not pushing this issue, just opening it for discussion.

baby-gnu replied:
Maybe `lookup=` is much more meanful than `namespace=`.

baby-gnu commented:
> @baby-gnu Just had a brief glance -- this looks like a very good idea. What break things -- just the commit message or something else?

The `files_switch` API is broken, I introduced a new **first parameter**.

To make it compatible with users calling directly `salt['config.get']` we should invert the two parameters like:

```
{%- macro files_switch(source_files,
                       namespace,
                       default_files_switch=['id', 'os_family'],
                       indent_width=6) %}
```

This way it can be used as 

```
    - source: {{ files_switch(
                    salt['config.get'](
                        tplroot ~ ':tofs:source_files:template-config-file-file-managed',
                        ['example.tmpl', 'example.tmpl.jinja']
                    )
              ) }}
```

or
 
```
    - source: {{ files_switch(['example.tmpl', 'example.tmpl.jinja'],
                              namespace='template-config-file-file-managed',
                 )
              }}
```

It could even be missused (doing the lookup two times) with:

```
    - source: {{ files_switch(
                    salt['config.get'](
                        tplroot ~ ':tofs:source_files:template-config-file-file-managed',
                        ['example.tmpl', 'example.tmpl.jinja']
                    ),
                    namespace='template-config-file-file-managed',

              ) }}
```

So, feature or refactor? ;-)

baby-gnu commented:
I rebased my work, integrating the different comments made.

The `default-debian-8-backports` test fail seems not related to my work.

I changed the new variable name, position and optionality.

People using the old format are not forced to change, they can still call `salt['config.get']` by themself:

```
    - source: {{ files_switch(
                    salt['config.get'](
                        tplroot ~ ':tofs:source_files:template-config-file-file-managed',
                        ['example.tmpl', 'example.tmpl.jinja']
                    )
                 )
              }}
```

or they can change and use the new optional `lookup` argument:

```
    - source: {{ files_switch(['example.tmpl', 'example.tmpl.jinja'],
                              lookup='template-config-file-file-managed'
                 )
              }}
```

There is no error if the formula writer mistakely added the `lookup` argument without changing the list of files:

```
    - source: {{ files_switch(
                    salt['config.get'](
                        tplroot ~ ':tofs:source_files:template-config-file-file-managed',
                        ['example.tmpl', 'example.tmpl.jinja']
                    ),
					lookup='template-config-file-file-managed'
                 )
              }}
```

Regards.

myii commented:
Restarted the failing tests and all OK from that end.

myii commented on template/macros.jinja:
```diff
@@ -7,6 +8,8 @@
 
     Params:
+      * lookup: key under '<tplroot>:tofs:source_files' to override
+        list of source files
```
_If_ we stick with this idea of using `lookup` as the second parameter, then this comment should be moved to under the `source_files` comment below.  The same order for both the parameters and the comments.

myii replied:
Thanks.

myii commented on template/macros.jinja:
```diff
@@ -52,6 +52,15 @@
   ) %}
+  {#- Lookup files or fallback to source_files parameter #}
+  {%- if not lookup %}
+    src_files = source_files
```
Need this `set` in a Jinja block.

myii replied:
But based on [the comment below](https://github.com/saltstack-formulas/template-formula/pull/77/files#r265373681), it will be removed anyway.

myii replied:
Thanks.

myii commented on template/macros.jinja:
```diff
@@ -52,6 +52,15 @@
+        tplroot ~ ':tofs:source_files:' ~ lookup,
+        source_files
+    ) %}
+  {%- endif %}
```
This whole `if-else-endif` block is unnecessary, since the `config.get` will result in the default value of `source_files` being used.  If there's any discomfort with `lookup=None` as the top, this can be set to any other default value.  But it appears to work fine as it is.

baby-gnu replied:
@myii thanks, I made some tests and you are right, the `salt['config.get']` is done with `template:tofs:source_files:None` which has no result so the second argument is used.

myii replied:
Thanks.

myii commented:
@baby-gnu I think this updated method is definitely an improvement -- thanks for the extra effort you've put in.  Please consider my new comments and let's wait for @aboe76 to review before merging.

myii commented:
@baby-gnu I'm really happy with this now -- an excellent refactor.  Thanks for your work and your patience.  Would you be willing to submit a follow-up PR to update [`docs/TOFS_pattern.rst`](https://github.com/saltstack-formulas/template-formula/blob/193bba805fc14d3f0f61c71b6ce4bbf45a60312c/docs/TOFS_pattern.rst)?

@aboe76 I'm happy with this to be merged, if you are OK with the changes as well.

baby-gnu commented:
> Would you be willing to submit a follow-up PR to update [`docs/TOFS_pattern.rst`](https://github.com/saltstack-formulas/template-formula/blob/193bba805fc14d3f0f61c71b6ce4bbf45a60312c/docs/TOFS_pattern.rst)?

I should have missed something @myii because I updated the example in `docs/TOFS_pattern.rst`.

myii commented:
@baby-gnu You're right, it has been updated!  I didn't look at the whole file, so I thought there may be more changes needed.

aboe76 commented:
@baby-gnu and @myii, you are both right, this looks like an improvement and I like the `lookup` part, this is a better pattern to use. LGTM

myii commented:
@babilen5 As discussed on IRC/Matrix, list of formulas to be updated:

Formula | Assignees | Status
---|---|---
[systemd-formula](https://github.com/saltstack-formulas/systemd-formula/pull/27)||Small update
[lighttpd-formula](https://github.com/saltstack-formulas/lighttpd-formula/pull/4)||Small update
[stack-formula](https://github.com/claudekenni/stack-formula/pull/2)|@myii|Small update
[zabbix-formula](https://github.com/saltstack-formulas/zabbix-formula)||From v1
[varnish-formula](https://github.com/saltstack-formulas/varnish-formula)||From v1

alxwr commented:
So `macros.jinja` now becomes a library which can be just dropped inside any formula? :-) I'm all for that. Maybe we want a `libtofs.jinja`? I really like the idea of being able to update standard functionality by copy-pasting the macro files I need into specific formulas. Copy-paste keeps the workload low while avoiding inter-formula dependencies.

baby-gnu commented:
Hello.

> So macros.jinja now becomes a library which can be just dropped inside any formula? :-) I'm all for that. Maybe we want a libtofs.jinja?

I like the idea, but for [systemd-formula](https://github.com/saltstack-formulas/systemd-formula) it looks like there are some specific things to lookup files under the sub-directories for each component.

I did not took enough time to look at it but I think a `libtofs.jinja` is a great idea since it meaningful.

Regards.

myii commented:
@alxwr I really like the idea of a `libtofs.jinja`, keeping macros in separate files for ease of reuse.  @baby-gnu Your concerns for the [systemd-formula](https://github.com/saltstack-formulas/systemd-formula) are well-founded -- I've got some progress to share, so I'm going to merge this PR in the meantime.  I'll put something up and ping you all soon.  Thanks for all of your work, reviews and ideas.

saltstack-formulas-travis commented:
:tada: This PR is included in version 1.2.4 :tada:

The release is available on [GitHub release](https://github.com/saltstack-formulas/template-formula/releases/tag/v1.2.4)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
